src/main/java/org/folio/edge/rtac/RtacHandler.java

```diff
@@ -39,76 +41,207 @@
+            if (checkAcceptHeadersByPattern(request, XML_RELATED_MIME_TYPE_PATTERN, false)) {
+              updateAcceptHeadersByPattern(request, XML_RELATED_MIME_TYPE_PATTERN, JSON_MIME_TYPE);
+              acceptHeadersUpdateFlag = true;
+            } else {
+              acceptHeadersUpdateFlag = false;
             }
-          })
-          .exceptionally(t -> {
-            logger.error("Exception calling mod-rtac", t);
-            returnEmptyResponse(ctx);
-            return null;
-          });
-      });
+          }
+
+          rtacClient.rtac(instanceIds, request.headers())
+              .thenAcceptAsync(body -> {
+                try {
+                  logger.debug("rtac response: {}", body);
+
+                  // Return back origin Accept headers if any
+                  if (acceptHeadersUpdateFlag) {
+                    updateAcceptHeadersWith(request, originAcceptHeaders);
+                  }
+
+                  final var isXmlResponse = checkAcceptHeadersXmlType(request);
+                  String returningContent = body;
+                  if (isXmlResponse) {
+                    final var instances = Instances.fromJson(body);
+                    if (isBatch) {
+                      returningContent = instances.toXml();
+                    } else {
+                      var holdings = instances.getHoldings().isEmpty() ? new Holdings()
+                          : instances.getHoldings().get(0);
+                      returningContent = holdings.toXml();
+                    }
+                  }
+
+                  logger.info("Converted Response: \n {}", returningContent);
+                  ctx.response()
+                      .setStatusCode(200)
+                      .putHeader(HttpHeaders.CONTENT_TYPE,
+                          isXmlResponse ? APPLICATION_XML : APPLICATION_JSON)
+                      .end(returningContent);
+
+                } catch (IOException e) {
+                  logger.error("Exception translating JSON -> XML: {}", e.getMessage(), e);
+                  returnEmptyResponse(ctx);
+                }
+              })
+              .exceptionally(t -> {
+                logger.error("Exception calling mod-rtac", t);
+                returnEmptyResponse(ctx);
+                return null;
+              });
+        });
+  }
+
+  /**
+   * EDGE-RTAC supports application/xml or application/json  and all its derivatives in Accept header.
+   * Empty Accept header implies any MIME type is accepted, same as Accept: *//*
+   *
+   * Valid examples: text/xml, text/*, *//*, *\xml, application/json, application/*, *//*, *\json
+   * NOT Valid examples: application/xml, application/*, test/json
+   *
+   * @param request - http request to the module
+   * @return - true if accept headers are supported
+   */
+  private boolean checkSupportedAcceptHeaders(HttpServerRequest request) {
```

Comment:
Can we extract this methods to a separate utility class so that they can be properly tested?

Reply:
Done.